src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +30,109 @@ public static class Renamer
                 newName, optionSet, cancellationToken);
         }
 
+        /// <summary>
+        /// Applies changes to the given document so that the folders, name, and path from <paramref name="newDocumentInfo"/>
+        /// are used. Will also perform symbol changes to the code in the document, attempting to sync type and namespace
+        /// to the new information.
+        /// </summary>
+        /// <returns><see cref="RenameDocumentResult"/> with the modified solution and indication of any symbol changes made</returns>
+        public static async Task<RenameDocumentResult> RenameDocumentAsync(
+            Document document, DocumentInfo newDocumentInfo, OptionSet optionSet, CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            var newName = newDocumentInfo.Name;
+            var shouldRenameType = !string.Equals(document.Name, newName, StringComparison.OrdinalIgnoreCase);
+            var shouldMoveNamespace = !document.Folders.SequenceEqual(newDocumentInfo.Folders);
+
+            var solution = document.Project.Solution
+                .WithDocumentName(document.Id, newDocumentInfo.Name)
+                .WithDocumentFilePath(document.Id, newDocumentInfo.FilePath)
+                .WithDocumentFolders(document.Id, newDocumentInfo.Folders);
+
+            if (!(shouldRenameType || shouldMoveNamespace))
+            {
+                return new RenameDocumentResult(solution, renameSymbolSucceeded: false);
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (shouldRenameType)
+            {
+                solution = await RenameTypeToMatchDocumentNameAsync(document.Id, solution, newName, document.Name, cancellationToken).ConfigureAwait(false);
+            }
+
+            if (shouldMoveNamespace)
+            {
+                solution = await SyncNamespaceToDocumentFoldersAsync(document.Id, solution, cancellationToken).ConfigureAwait(false);
+            }
+
+            return new RenameDocumentResult(solution, renameSymbolSucceeded: true);
+        }
+
+        private static async Task<Solution> RenameTypeToMatchDocumentNameAsync(DocumentId documentId, Solution solution, string newName, string originalName, CancellationToken cancellationToken)
+        {
+            var newNameWithoutExtension = Path.GetFileNameWithoutExtension(newName);
```

Comment:
1. i think we need to really validate that newNameWithoutExtension is legal
2. what happens if i have `Foo.Bar.cs` and i rename that to `Foo.Baz.cs`?

Reply:
`Foo.Bar.cs` to `Foo.Baz.cs` technically would be fine, but that's only because no type name would be `Foo.Bar` :P 

I get your point though. My assumption was that validating with the `RenameSymbolAsync` API would be enough. Maybe not though... it might be more useful to check here and provide details on why the type won't be renamed if that happens.

Reply:
Well, types absolutely could be named Foo.Bar if this is a nested type like we have in this PR itself. :smile:

Reply:
no _root_ type (parent is a namespace?) would be named that. We could potentially beef this logic up to handle common partial/nested types. I'd prefer to have that as a follow up item and not block this.

Reply:
Yeah don't block on this, but it seems we Eventually™ need some fancier logic around a part of a file name changing and realizing that's a nested/partial class and change just the right bit.